sys/linux/fs_ioctl_btrfs.txt

```diff
@@ -1,7 +1,426 @@
+
+#add a padding since fd in this struct should be 64 bits (is this the correct way?)
+btrfs_ioctl_vol_args {
+	fd	fd
+	dummy	const[0, int32]
+	name	array[int8, BTRFS_PATH_MAX]
+}
+
+btrfs_ioctl_vol_args_v2 {
+	fd	fd
+	transid	transid
+	flags	flags[btrfs_ioctl_vol_args_v2_flags, int64]
+	u1	btrfs_args_v2_u1
+	u2	btrfs_args_v2_u2
+}
+
+btrfs_args_v2_u1 [
+	inherit	btrfs_args_v2_u1_s1
+	unused	array[int64, 4]
+]
+
+btrfs_args_v2_u2 [
+	name		array[int8, BTRFS_PATH_MAX]
+	devid		devid
+	subvolid	int64
+]
+
+btrfs_args_v2_u1_s1 {
+	size		bytesize[qgroup_inherit, int64]
+	qgroup_inherit	ptr[in, btrfs_qgroup_inherit]
+}
+
+btrfs_qgroup_inherit {
+	flags		flags[btrfs_qgroup_inherit_flags, int64]
+	num_qgroups	int64
+	num_ref_copies	int64
+	num_excl_copies	int64
+	lim		btrfs_qgroup_limit
+}
+
+btrfs_qgroup_limit {
+	flags		flags[btrfs_qgroup_limit_flags, int64]
+	max_rfer	int64
+	max_excl	int64
+	rsv_rfer	int64
+	rsv_excl	int64
+}
+
+btrfs_ioctl_defrag_range_args {
+	start		int64
+	len		int64
+	flags		flags[btrfs_ioctl_defrag_range_args_flags, int64]
+	extent_thresh	int32
+	compress_type	flags[btrfs_compression_type, int32]
+	unused		array[int32, 4]
+}
+
+btrfs_ioctl_fs_info_args {
+	max_id		int64
+	num_devices	int64
+	fsid		array[int8, BTRFS_FSID_SIZE]
+	nodesize	int32
+	sectorize	int32
+	clone_alignment	int32
+	reserved32	int32
+	reserverxd	array[int64, 122]
+}
+
+btrfs_ioctl_dev_info_args {
+	devid		devid
+	uuid		array[int8, BTRFS_UUID_SIZE]
+	byres_used	int64
+	total_bytes	int64
+	unused		array[int64, 379]
+	path		array[int8, BTRFS_DEVICE_PATH_NAME_MAX]
+}
+
+btrfs_ioctl_search_args {
+	key	btrfs_ioctl_search_key
+	buf	array[int8, BTRFS_SEARCH_ARGS_BUFSIZE]
+}
+
+btrfs_ioctl_search_key {
+	tree_id		treeid
+	min_objectid	int64
+	max_objectid	int64
+	min_offset	int64
+	max_offset	int64
+	min_transid	int64
+	max_transid	int64
+	min_type	int32
+	max_type	int32
+	nr_items	int32
+	unused		int32
+	unused1		int64
+	unused2		int64
+	unused3		int64
+	unused4		int64
+}
+
+#it seems that there is no limit for the length of buf?
+btrfs_ioctl_search_args_v2 {
+	key		btrfs_ioctl_search_key
+	buf_size	len[buf, int64]
+	buf		array[int64]
+}
+
+btrfs_ioctl_ino_lookup_args {
+	treeid		treeid
+	objectid	int64
+	name		array[int8, BTRFS_INO_LOOKUP_PATH_MAX]
+}
+
+btrfs_ioctl_ino_path_args {
+	inum		int64
+	size		int64
+	reserved	array[int64, 4]
+	fspath		int64
+}
+
+btrfs_ioctl_logical_ino_args {
+	logical		int64
+	size		int64
+	reserved	array[const[0, int64], 3]
+	flags		flags[btrfs_ioctl_logical_ino_args_flags, int64]
+	inodes		int64
+}
+
+btrfs_ioctl_space_args {
+	space_slots	len[spaces, int64]
+	total_spaces	int64
+	spaces		array[btrfs_ioctl_space_info, 0:PAGE_SIZE]
+}
+
+btrfs_ioctl_space_info {
+	flags		int64
+	total_bytes	int64
+	used_bytes	int64
+}
+
+# btrfs_scrub_progress is ignored since that will only be used for out
+btrfs_ioctl_scrub_args {
+	devid	devid
+	start	int64
+	end	int64
+	flags	flags[btrfs_ioctl_scrub_args_flags, int64]
+	unused	array[int8, 992]
```

Comment:
Changed, though I'm really curious what the difference would be (i.e. why would it not correctly aligned if we write it as int8?)

Reply:
In this particular case, it won't matter because the previous field is int64. But generally it affects layout.
Consider:
```
struct foo {
  char a;
  char b[8];
};
```
vs:
```
struct bar {
  char a;
  int64 b;
};
```
foo.b starts at offset 1, while bar.b starts at offset 8 (with 7 bytes of padding before it). Also overall struct layout is different: 1 vs 8. So it may affect how foo/bar is allocated within outer struct as well.